src/client/executor.rs

```diff
@@ -530,82 +523,79 @@ impl Client {
         let start_time = Instant::now();
         let command_result = match connection.send_raw_command(raw_cmd, request_id).await {
             Ok(response) => {
-                match T::Response::deserialize_response(&response) {
-                    Ok(r) => {
-                        if let (Some(session), Some(ts)) = (session.as_mut(), r.operation_time()) {
-                            session.advance_operation_time(ts);
-                        }
-                        self.update_cluster_time(&r, session).await;
-                        if r.is_success() {
-                            // Retrieve recovery token from successful response.
-                            Client::update_recovery_token(is_sharded, &r, session).await;
-
-                            Ok(CommandResult {
-                                raw: response,
-                                deserialized: r.into_body(),
-                            })
-                        } else {
-                            // if command was ok: 0, try to deserialize the command error.
-                            // if that fails, return a generic error.
-                            Err(response
-                                .body::<CommandErrorBody>()
-                                .map(|error_response| error_response.into())
-                                .unwrap_or_else(|e| {
-                                    Error::from(ErrorKind::InvalidResponse {
-                                        message: format!(
-                                            "error deserializing command error: {}",
-                                            e
-                                        ),
-                                    })
-                                }))
+                async fn handle_response<T: Operation>(
+                    client: &Client,
+                    op: &T,
+                    session: &mut Option<&mut ClientSession>,
+                    is_sharded: bool,
+                    response: RawCommandResponse,
+                ) -> Result<RawCommandResponse> {
+                    let raw_doc = RawDocument::new(response.as_bytes())?;
+
+                    let ok = match raw_doc.get("ok")? {
```

Comment:
doing linear scans of the command result for each of these keys shouldn't be a performance concern, since the top level of a command response rarely has many keys to sift through.